storage/innobase/row/row0upd.cc

```diff
@@ -420,10 +420,8 @@ wsrep_row_upd_check_foreign_constraints(
 		NOT break the constraint. */
 
 		if (foreign->foreign_index == index
-		    && (node->is_delete
-			|| row_upd_changes_first_fields_binary(
-				entry, index, node->update,
-				foreign->n_fields))) {
+		    && row_upd_changes_first_fields_binary(
```

Comment:
I'm trying to understand why node->is_delete is not required here. I can see that on `row_upd_sec_index_entry` we call `row_upd_index_is_referenced` which is the base function for `wsrep_row_upd_check_foreign_constraints`, then if `WITH_WSREP` is defined we call `wsrep_row_upd_check_foreign_constraints` which basically runs a check from `row_ins_check_foreign_constraint` on the parent table instead of on the child table (calling `row_ins_check_foreign_constraint` with 1st parameter = TRUE). I have two questions:

1. Why this second call to check FK constrain is required only on WSREP
2. By removing `node->is_delete` we won't be calling the function that is causing the issue, what happens on an update that returns true for  row_upd_changes_first_fields_binary and has the parent table records locked?

Reply:
While talking with @venkatesh-prasad-v about this fix we found out that test case is failing for UPDATES:

```
setup
CREATE TABLE parent_t ( id INT AUTO_INCREMENT PRIMARY KEY );
CREATE TABLE child_t ( id INT AUTO_INCREMENT PRIMARY KEY, parent_id INT,
CONSTRAINT `FK_child_parent_id` FOREIGN KEY (`parent_id`) REFERENCES `parent_t` (`id`));
INSERT INTO parent_t VALUES (), (), (), (), ();
INSERT INTO child_t (parent_id) SELECT id FROM parent_t;
Session 1
BEGIN;
SELECT * FROM parent_t WHERE id = 1 FOR UPDATE;
Session 2
UPDATE child_t SET parent_id = 2 WHERE parent_id = 1;
Session 1
COMMIT;--(ROLLBACK;, abort, doesn't matter)
Session 2
ERROR 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
```

Reply:
Hi Marcelo,

Thank you for your time on reviewing this.

> 1. Why this second call to check FK constrain is required only on WSREP

The call to `row_upd_index_is_referenced()` function is for checking if there are any references to the index in the current table. i.e in simple words, it only means that if there are any child tables referring to the current table object.

Whereas the the call to `wsrep_row_upd_index_is_foreign()` function is for checking if the current table has a any foreign key defined.

Both these calls required since they will is used by the `wsrep_append_foreign_key` for generating writesets for the update operation.

> 2. By removing `node->is_delete` we won't be calling the function that is causing the issue, what happens on an update that returns true for row_upd_changes_first_fields_binary and has the parent table records locked?

In such cases, updates also wait for the lock on the parent row to be released. However there was an issue in the error handling code that made the UPDATE queries to fail even after the parent rows were released. This has been fixed in the latest version of the patch.